### PR TITLE
Add Blob AzureWebApp provider

### PR DESCRIPTION
**Pull request by pakrym**

### CODE REVIEW

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                    eventGroup.Key.Item4, BlobPathSeparator,
+                    _fileName
+                );
+
+                var blob = _container.GetAppendBlobReference(blobName);
+
+                CloudBlobStream stream;
+                try
+                {
+                    stream = await blob.OpenWriteAsync(false);
```
Comment (davidfowl) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs]: false? What's the argument name

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsLoggerProvider.cs [hunk 2]
```diff
@@ -13,45 +14,50 @@ public class AzureWebAppDiagnosticsLoggerProvider : ILoggerProvider
...
         {
-            return _innerLoggerProvider.CreateLogger(categoryName);
+            return _loggerFactory?.CreateLogger(categoryName) ?? NullLogger.Instance;
         }
 
         /// <inheritdoc />
         public void Dispose()
         {
-            _innerLoggerProvider.Dispose();
+            _loggerFactory.Dispose();
```
Comment (campersau) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsLoggerProvider.cs]: I think this can also be `null`.

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsFactoryExtensions.cs [hunk 3]
```diff
@@ -15,13 +15,22 @@ public static class AzureWebAppDiagnosticsFactoryExtensions
         /// Adds an Azure Web Apps diagnostics logger.
         /// </summary>
         /// <param name="factory">The extension method argument</param>
-        /// <param name="fileSizeLimitMb">A strictly positive value representing the maximum log size in megabytes. Once the log is full, no more message will be appended</param>
-        public static ILoggerFactory AddAzureWebAppDiagnostics(this ILoggerFactory factory, int fileSizeLimitMb = FileLoggerProvider.DefaultFileSizeLimitMb)
+        public static ILoggerFactory AddAzureWebAppDiagnostics(this ILoggerFactory factory)
+        {
+            return AddAzureWebAppDiagnostics(factory, null);
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsFactoryExtensions.cs]: null = ?

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsSettings.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace Microsoft.Extensions.Logging.AzureWebAppDiagnostics
+{
+    /// <summary>
+    /// Settings for <see cref="AzureWebAppDiagnosticsLoggerProvider"/>.
+    /// </summary>
+    public class AzureWebAppDiagnosticsSettings
+    {
+        /// <summary>
+        /// A strictly positive value representing the maximum log size in bytes. Once the log is full, no more message will be appended
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsSettings.cs]: The standard message for properties that are get & set starts with "Gets or sets ..."

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+        /// <inheritdoc />
+        protected override async Task EmitBatchAsync(IEnumerable<LogEvent> events)
+        {
+            var eventGroups = events.GroupBy(GetBlobKey);
+            foreach (var eventGroup in eventGroups)
+            {
+                var blobName = string.Concat(
+                    _appName, BlobPathSeparator,
+                    eventGroup.Key.Item1, BlobPathSeparator,
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs]: Add some comments and examples about Item1 ... Item4. What do they mean?

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                    eventGroup.Key.Item4, BlobPathSeparator,
+                    _fileName
+                );
+
+                var blob = _container.GetAppendBlobReference(blobName);
+
+                CloudBlobStream stream;
+                try
+                {
+                    stream = await blob.OpenWriteAsync(createNew: false);
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs]: How will this work in a multi instance scenario? Do all instances of the website write in the same blob? Will it conflict?

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/WebConfigurationReaderLevelSwitch.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        private static readonly LogEventLevel LogLevelDisabled = LogEventLevel.Fatal + 1;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="WebConfigurationReaderLevelSwitch"/> class.
+        /// </summary>
+        /// <param name="reader"></param>
+        /// <param name="convert"></param>
+        public WebConfigurationReaderLevelSwitch(IWebAppLogConfigurationReader reader, Func<WebAppLogConfiguration, LogLevel> convert )
+        {
+            reader.OnConfigurationChanged += (sender, configuration) =>
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/WebConfigurationReaderLevelSwitch.cs]: Potential for leaking event handlers here. You never unsubscribe

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobLoggerProvider.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        /// Creates a new instance of the <see cref="AzureBlobLoggerProvider"/> class.
+        /// </summary>
+        /// <param name="outputTemplate"></param>
+        /// <param name="appName"></param>
+        /// <param name="fileName"></param>
+        /// <param name="batchSize"></param>
+        /// <param name="period"></param>
+        public AzureBlobLoggerProvider(string outputTemplate, string appName, string fileName, int batchSize, TimeSpan period)
+        {
+            _outputTemplate = outputTemplate;
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobLoggerProvider.cs]: Any checks for these parameters? You'll probably gets some null refs later if not

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            ITextFormatter formatter,
+            int batchSizeLimit,
+            TimeSpan period) : base(batchSizeLimit, period)
+        {
+            _appName = appName;
+            _fileName = fileName;
+            _formatter = formatter;
+            if (batchSizeLimit < 1)
+            {
+                throw new ArgumentException(nameof(batchSizeLimit));
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs]: Argument out of range exception. Also add a text explaining to the poor user what they did wrong

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/FileLoggerProvider.cs [hunk 10]
```diff
@@ -4,69 +4,65 @@
...
-
-                var fileSizeLimitBytes = fileSizeLimit * 1024 * 1024;
-
-                var messageFormatter = new MessageTemplateTextFormatter(OutputTemplate, null);
-                var rollingFileSink = new RollingFileSink(logsFilePattern, messageFormatter, fileSizeLimitBytes, RetainedFileCountLimit);
-                var backgroundSink = new BackgroundSink(rollingFileSink, BackgroundSink.DefaultLogMessagesQueueSize);
-
-                loggerConfiguration.WriteTo.Sink(backgroundSink);
-            })
+        /// <param name="retainedFileCountLimit"></param>
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/FileLoggerProvider.cs]: I'm not sure that we want to make this configurable yet. Also, by changing this, you need to at least add some checks to validate all those comments on the properties (strictly positive, etc.)

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/SerilogLoggerProvider.cs [hunk 11]
```diff
@@ -4,122 +4,19 @@
...
 using Serilog.Core;
-using Serilog.Events;
 
 namespace Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Internal
 {
     /// <summary>
     /// Represents a Serilog logger provider use for Azure WebApp.
     /// </summary>
-    public abstract class SerilogLoggerProvider : ILoggerProvider
+    public abstract class SerilogLoggerProvider
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/SerilogLoggerProvider.cs]: Do you need this class any more?

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/WebConfigurationReaderLevelSwitch.cs (hunk 7 above)
Comment (pakrym) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/WebConfigurationReaderLevelSwitch.cs]: We never replace or recreate loggers so their lifetime is similar to `IWebAppLogConfigurationReader` lifetime so I think handler leak is not an issue here.

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs (hunk 6 above)
Comment (pakrym) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs]: Yes, Azure documentation states that it's safe for multiple writers, it will append in `4MB` blocks by default.

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs (hunk 5 above)
Comment (pakrym) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs]: They are created one method down and used just to group event, do you feel this needs more comments?

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs (hunk 6 above)
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs]: Then should you write some sort of instance ID with the messages? Otherwise, it will be quite hard to figure out which instance owns the messages

**Comment by victorhurdugaci** [comment]
You didn't compile locally after your last changes, didn't you? 😏  (travis/appveyor)

**Comment by pakrym** [comment]
Don't tell anybody :) Also thank you for reviewing while on vacation.

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs (hunk 6 above)
Comment (pakrym) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs]: Now I think It's better to include instance id into blob name.

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobLoggerProvider.cs [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+        /// <inheritdoc />
+        public Logger ConfigureLogger(IWebAppLogConfigurationReader reader)
+        {
+            var messageFormatter = new MessageTemplateTextFormatter(_outputTemplate, null);
+            var container = new CloudBlobContainer(new Uri(reader.Current.BlobContainerUrl));
+            var fileName = _instanceId + "-" + _fileName;
+            var azureBlobSink = new AzureBlobSink(container, _appName, fileName, messageFormatter, _batchSize, _period);
+            var backgroundSink = new BackgroundSink(azureBlobSink, BackgroundSink.DefaultLogMessagesQueueSize);
+            LoggerConfiguration loggerConfiguration = new LoggerConfiguration();
```
Comment (cesarbs) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobLoggerProvider.cs]: nit: `var`

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobLoggerProvider.cs [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+            loggerConfiguration.WriteTo.Sink(backgroundSink);
+            loggerConfiguration.MinimumLevel.ControlledBy(new WebConfigurationReaderLevelSwitch(reader,
+                configuration =>
+                {
+                    return configuration.BlobLoggingEnabled ? configuration.BlobLoggingLevel: LogLevel.None;
```
Comment (cesarbs) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobLoggerProvider.cs]: nit: add a space before the `:`

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsSettings.cs [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace Microsoft.Extensions.Logging.AzureWebAppDiagnostics
+{
+    /// <summary>
+    /// Settings for <see cref="AzureWebAppDiagnosticsLoggerProvider"/>.
+    /// </summary>
+    public class AzureWebAppDiagnosticsSettings
+    {
+        /// <summary>
+        /// Gets or sets a strictly positive value representing the maximum log size in bytes. Once the log is full, no more message will be appended
```
Comment (cesarbs) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsSettings.cs]: nit: finish with period.

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsSettings.cs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        /// </summary>
+        public int FileSizeLimit { get; set; } = 10 * 1024 * 1024;
+
+        /// <summary>
+        /// Gets or sets a strictly positive value representing the maximum retained file count
```
Comment (cesarbs) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsSettings.cs]: period

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsSettings.cs [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+        /// </summary>
+        public int RetainedFileCountLimit { get; set; } = 2;
+
+        /// <summary>
+        /// Gets or sets a message template describing the output messages
```
Comment (cesarbs) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsSettings.cs]: period

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsSettings.cs [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+        /// </summary>
+        public string OutputTemplate { get; set; } = "{Timestamp:yyyy-MM-dd HH:mm:ss.fff zzz} [{Level}] {Message}{NewLine}{Exception}";
+
+        /// <summary>
+        /// Gets or sets a maximum number of events to include in a single blob append batch.
+        /// </summary>
+        public int BlobBatchSize { get; set; } = 32;
+
+        /// <summary>
+        /// Gets or sets a time to wait between checking for blob log batches
```
Comment (cesarbs) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsSettings.cs]: period

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/FileLoggerProvider.cs [hunk 18]
```diff
@@ -4,69 +4,78 @@
...
-                LevelSwitcher.MinimumLevel = LogLevelToLogEventLevel(newConfiguration.FileLoggingLevel);
+                Directory.CreateDirectory(logsFolder);
             }
+            var logsFilePattern = Path.Combine(logsFolder, FileNamePattern);
+
+            var messageFormatter = new MessageTemplateTextFormatter(_outputTemplate, null);
+            var rollingFileSink = new RollingFileSink(logsFilePattern, messageFormatter, _fileSizeLimit, _retainedFileCountLimit);
+            var backgroundSink = new BackgroundSink(rollingFileSink, BackgroundSink.DefaultLogMessagesQueueSize);
+
+            LoggerConfiguration loggerConfiguration = new LoggerConfiguration();
```
Comment (cesarbs) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/FileLoggerProvider.cs]: `var`

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/WebConfigurationReaderLevelSwitch.cs [hunk 19]
```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
```
Comment (cesarbs) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/WebConfigurationReaderLevelSwitch.cs]: nit: add a blank line after the copyright comment

File: test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/SerilogLoggerProviderTests.cs [hunk 20]
```diff
@@ -140,68 +134,69 @@ public void LevelMapping()
...
+                currentConfig = configBuilder.Build();
+                configReader.Raise(m => m.OnConfigurationChanged += (sender, e) => { }, null, currentConfig);
 
-                    // Don't try to log "None" (start at 1)
-                    for (int j = 1; j < levelsToCheck.Length; j++)
-                    {
-                        logger.Log(levelsToCheck[j], 1, new object(), null, (state, ex) => string.Empty);
-                    }
+                // Don't try to log "None" (start at 1)
+                for (int j = 1; j < levelsToCheck.Length; j++)
```
Comment (cesarbs) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/SerilogLoggerProviderTests.cs]: `var`

File: test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/SerilogLoggerProviderTests.cs [hunk 21]
```diff
@@ -140,68 +134,69 @@ public void LevelMapping()
...
+                LogLevel.Debug,
+                LogLevel.Trace
+            };
+
+            var seriloglogger = provider.ConfigureLogger(configReader.Object);
+            var loggerFactory = new LoggerFactory();
+            loggerFactory.AddSerilog(seriloglogger);
+            var logger = loggerFactory.CreateLogger("TestLogger");
+
+            for (int i = 0; i < levelsToCheck.Length; i++)
```
Comment (cesarbs) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/SerilogLoggerProviderTests.cs]: `var`

File: test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/SerilogLoggerProviderTests.cs [hunk 22]
```diff
@@ -140,68 +134,69 @@ public void LevelMapping()
             var testSink = new Mock<ILogEventSink>(MockBehavior.Strict);
             testSink.Setup(m => m.Emit(It.IsAny<LogEvent>()));
 
-            using (var provider = new TestWebAppSerilogLoggerProvider(testSink.Object, configReader.Object))
+            var provider = new TestWebAppSerilogLoggerProvider(testSink.Object);
+            var levelsToCheck = new LogLevel[]
```
Comment (cesarbs) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/SerilogLoggerProviderTests.cs]: I think you don't need the type here i.e. just use `new []`

File: test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/SerilogLoggerProviderTests.cs [hunk 23]
```diff
@@ -105,31 +101,29 @@ public void DynamicLoggingLevel()
...
-                testSink.Verify(m => m.Emit(It.IsAny<LogEvent>()), Times.Once);
-            }
+            // Logging for this level should be enabled now
+            logger.Debug("Test1");
+            testSink.Verify(m => m.Emit(It.IsAny<LogEvent>()), Times.Once);
         }
 
         // Checks that the .net log level to serilog level mappings are doing what we expect
         [Fact]
         public void LevelMapping()
```
Comment (cesarbs) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/SerilogLoggerProviderTests.cs]: Instead of the loops in this test, couldn't you turn it into a `Theory` and generate the data for each `i,j` test case?

File: test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/project.json [hunk 24]
```diff
@@ -17,7 +17,8 @@
           "version": "1.0.0-*",
           "type": "platform"
         }
-      }
+      },
+      "imports": [ "portable-net40+sl5+win8+wp8+wpa81", "portable-net45+win8+wp8+wpa81" ]
```
Comment (cesarbs) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/project.json]: why is this needed?

Comment (pakrym) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/project.json]: Storage APIs require imports.

**Comment by pakrym** [comment]
:watch:

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsLoggerProvider.cs [hunk 25]
```diff
@@ -13,45 +14,51 @@ public class AzureWebAppDiagnosticsLoggerProvider : ILoggerProvider
...
 
-        private readonly ILoggerProvider _innerLoggerProvider;
-        private readonly bool _runningInWebApp;
+        private readonly LoggerFactory _loggerFactory;
 
         /// <summary>
         /// Creates a new instance of the <see cref="AzureWebAppDiagnosticsLoggerProvider"/> class.
         /// </summary>
-        public AzureWebAppDiagnosticsLoggerProvider(WebAppContext context, int fileSizeLimitMb)
+        public AzureWebAppDiagnosticsLoggerProvider(WebAppContext context, AzureWebAppDiagnosticsSettings settings)
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/AzureWebAppDiagnosticsLoggerProvider.cs]: Null check for the new arg. You'll get a null ref otherwise when you create the FileLoggerProvider

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobLoggerProvider.cs [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                throw new ArgumentNullException(nameof(fileName));
+            }
+            if (batchSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(batchSize), $"{nameof(batchSize)} should be a positive number.");
+            }
+            if (period <= TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(period), $"{nameof(period)} should be longer then zero.");
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobLoggerProvider.cs]: why should the time space be longer and then it should be zero? :trollface:  (thAn)

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobLoggerProvider.cs [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+        }
+
+        /// <inheritdoc />
+        public Logger ConfigureLogger(IWebAppLogConfigurationReader reader)
+        {
+            var messageFormatter = new MessageTemplateTextFormatter(_outputTemplate, null);
+            var container = new CloudBlobContainer(new Uri(reader.Current.BlobContainerUrl));
+            var fileName = _instanceId + "-" + _fileName;
+            var azureBlobSink = new AzureBlobSink(container, _appName, fileName, messageFormatter, _batchSize, _period);
+            var backgroundSink = new BackgroundSink(azureBlobSink, BackgroundSink.DefaultLogMessagesQueueSize);
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobLoggerProvider.cs]: I think the queue size should be configurable

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+            {
+                throw new ArgumentNullException(nameof(formatter));
+            }
+            if (batchSizeLimit <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(batchSizeLimit), $"{nameof(batchSizeLimit)} should be a positive number.");
+            }
+            if (period <= TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(period), $"{nameof(period)} should be longer then zero.");
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/AzureBlobSink.cs]: thAn

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/FileLoggerProvider.cs [hunk 29]
```diff
@@ -4,69 +4,78 @@
...
 using Serilog;
+using Serilog.Core;
 using Serilog.Formatting.Display;
 using Serilog.Sinks.RollingFile;
 
 namespace Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Internal
 {
     /// <summary>
-    /// A file logger for Azure WebApp.
+    /// The logger provider that creates instances of <see cref="Serilog.Core.Logger"/> connected to <see cref="RollingFileSink"/>.
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/FileLoggerProvider.cs]: "Connected to..." is not needed. That's an implementation detail

File: src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/FileLoggerProvider.cs [hunk 30]
```diff
@@ -4,69 +4,78 @@
...
             }
-            else
+
+            var logsFolder = webAppConfiguration.FileLoggingFolder;
+            if (!Directory.Exists(logsFolder))
             {
-                LevelSwitcher.MinimumLevel = LogLevelToLogEventLevel(newConfiguration.FileLoggingLevel);
+                Directory.CreateDirectory(logsFolder);
             }
+            var logsFilePattern = Path.Combine(logsFolder, FileNamePattern);
```
Comment (victorhurdugaci) [inline src/Microsoft.Extensions.Logging.AzureWebAppDiagnostics/Internal/FileLoggerProvider.cs]: Probably this should include the instance id too because the network share is shared between instances

**Comment by victorhurdugaci** [comment]
The code looks okay but it would be nice to have some tests that actually check the azure sdk part

**Comment by pakrym** [comment]
I tries to write tests but it's nearly impossible to mock out azure blob types.

**Comment by victorhurdugaci** [comment]
You could try to use the emulator for tests or you might want to take a look at [WebJobs](https://github.com/Azure/azure-webjobs-sdk). I remember there was some storage mocking there. If you get any code from there, make sure you sync with Eilon first - he was the manager for that project.

**Comment by pakrym** [comment]
@victorhurdugaci added tests.

File: test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+            var buffers = new List<byte[]>();
+            blob.Setup(b => b.OpenWriteAsync()).Returns(() => Task.FromResult((Stream)new TestMemoryStream(buffers)));
+
+            var logger = CreateLogger(name => blob.Object);
+
+            for (int i = 0; i < 5; i++)
+            {
+                logger.Information("Text " + i);
+            }
+            await Task.Delay(TimeSpan.FromSeconds(1));
```
Comment (victorhurdugaci) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs]: no, no... use some sort of polling. Hardcoded wait times are evil

File: test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs [hunk 32]
```diff
@@ -0,0 +1,157 @@
...
+#if NET451
+            Assert.Equal(1, buffers.Count);
+            Assert.Equal(Encoding.UTF8.GetString(buffers[0]), @"Information Text 0
+Information Text 1
+Information Text 2
+Information Text 3
+Information Text 4
+");
+#else
+            // PeriodicBatchingSink always writes first message as seperate batch on coreclr
```
Comment (victorhurdugaci) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs]: Bug? Should be reported?

File: test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs [hunk 33]
```diff
@@ -0,0 +1,157 @@
...
+            var logger = CreateLogger(name =>
+            {
+                names.Add(name);
+                return blob.Object;
+            });
+
+            for (int i = 0; i < 3; i++)
+            {
+                logger.Write(new LogEvent(
+                    new DateTimeOffset(new DateTime(2016, 8, 29, i, 0, 0)),
```
Comment (victorhurdugaci) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs]: If you want to make this test cover everything, you should check also the boundaries outside the interval. Make sure that right after this, it's a new day and before is the same

File: test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs (hunk 31 above)
Comment (victorhurdugaci) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs]: You can probably poll the buffer until you get the expected number of messages. (with a timeout, of course)

**Comment by pakrym** [comment]
@victorhurdugaci switched to `CountdownEvent`

File: test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs (hunk 32 above)
Comment (pakrym) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs]: https://github.com/serilog/serilog-sinks-periodicbatching/issues/7

File: test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs [hunk 34]
```diff
@@ -0,0 +1,179 @@
...
+
+            var sink = new TestAzureBlobSink(name => blob.Object, 5);
+            var logger = CreateLogger(sink);
+
+            for (int i = 0; i < 5; i++)
+            {
+                logger.Information("Text " + i);
+            }
+
+            Assert.True(sink.CountdownEvent.Wait(TimeSpan.FromSeconds(1)));
```
Comment (victorhurdugaci) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs]: Be more reasonable. Give it 10 seconds

Comment (pakrym) [inline test/Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test/AzureBlobSinkTests.cs]: Done.

**Comment by victorhurdugaci** [comment]
:shipit: once Travis and AppVeyor pass

**Comment by pakrym** [comment]
@victorhurdugaci  `appveyor/pr` fails with 

```
Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test.LogConfigurationReaderTests.ConfigurationChange [FAIL]
      System.TimeoutException : Maximum number of retries reached.
      Stack Trace:
        C:\projects\logging\test\Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test\RetryHelper.cs(25,0): at Microsoft.Extensions.Logging.RetryHelper.Retry(Func`1 action, Int32 retries)
        C:\projects\logging\test\Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test\LogConfigurationReaderTests.cs(179,0): at Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test.LogConfigurationReaderTests.ConfigurationChange()
  Finished:    Microsoft.Extensions.Logging.AzureWebAppDiagnostics.Test
```

Which is unrelated to this PR. Merging.
